Add help files for all fields

proski opened the pull request:
I don't consider it a release blocker, but I think it's a matter of courtesy to the users. I believe users would be more willing to report issues when the observed behavior mismatches the documentation.

The changes include:

* Help files for all UI fields
* Improvements for some labels (consistent capitalization, avoiding abbreviations)
* README.md updated to reflect the updated label names, expanded explanations, uniform formatting.

proski committed:
Add help files for all fields

jakub-bochenski commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-ciBuildPhrases.html:
```diff
@@ -0,0 +1,6 @@
+<div>
+    Build the PR if the specified phrase has been posted as a PR comment
+    in Stash, even if the PR has not changed.<br/>
```
Actually is it a regex or literal match? Not sure myself, would have to check the code

jakub-bochenski replied:
Actually it's a case-insensitive substring match https://github.com/jenkinsci/stash-pullrequest-builder-plugin/blob/master/src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java#L396

proski replied:
I confirm it. But I'm not sure we should document implementation details if chances are we would change them.

jakub-bochenski replied:
I would rather explain the current behavior.

Changing this properly should involve a data migration anyway, it can be used to convert the old value to a regex

jakub-bochenski replied:
@proski so what about this? I'm not insisting, just making sure you didn't miss the comment above

proski replied:
Done

jakub-bochenski commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-stashHost.html:
```diff
@@ -0,0 +1,4 @@
+<div>
+    The base URL of the Stash server REST API.
+    This is <strong>not</strong> a URL for git.
```
I think this might still be ambiguous. An example would be useful, also I'd phrase this like `Note: this is not the git clone URL`.
Please see the description in readme file.

proski replied:
Adding more emphasis.

proski committed:
Adjust the comments

mergeOnSuccess is fortunately smart enough to only merge matrix builds
that succeed in all configurations.

Emphasize even stronger that Stash URL is not a git URL.

proski committed:
Improve field labels, fix capitalization

Always call "test this please" the build phrase. Avoid the "CI"
abbreviation.

proski commented:
I'm very happy that my assumption about `mergeOnSuccess` was wrong. stash-pullrequest-builder-plugin could be an example for other plugins how to do something in the top-level job without depending on the matrix plugin.

I've improved the labels as well, mostly capitalization.

`README.md` should be adjusted to match. I'm not sure if we need to document every option there now that we have online help. I think we can just list the important features. That would be a separate change.

I still have some doubts about `cancelOutdatedJobsEnabled`. I was able to reproduce removing outdated jobs from the queue, which is great. But I wasn't able to reproduce aborting the running jobs. Please look for `abortRunningJobsThatMatch` in the code. I don't know when it would actually activate. I'd rather not have that feature unless it's very well documented.

jakub-bochenski commented:
> README.md should be adjusted to match. I'm not sure if we need to document every option there now that we have online help. I think we can just list the important features. 

Yeah, I'm worried about maintaining those duplicate descriptions. Consider there is also another version on the plugin wiki at https://wiki.jenkins.io/display/JENKINS/Stash+pullrequest+builder+plugin

One way to solve this would be to just remove the descriptions from readme and the wiki. The problem with this approach is that plugin help is not easily accessible if you don't have it installed.

proski commented:
OK, I'll make changes to `README.md` to make the descriptions match the field labels. The Wiki can be edited later, it should reflect the latest release.

jakub-bochenski commented:
I was thinking about just removing the wiki description completely and just having a link to readme there instead. Thats one less place to keep updated. WDYT?

proski commented:
> I was thinking about just removing the wiki description completely and just having a link to readme there instead. Thats one less place to keep updated. WDYT?

I agree, that would be easier to maintain.

proski committed:
Update README.md to match labels and descriptions

Use uniform formatting, make all labels bold.

Minor adjustments to labels and descriptions.

proski commented:
I've updated `README.md`. I'm not doing force commits to facilitate incremental reviews, but please don't hesitate to merge the commits when merging this PR.

proski committed:
Add top-level help files for StashBuildTrigger and StashPostBuildComment

proski committed:
Avoid title case in labels, use "Stash pull request builder" uniformly

proski committed:
Explained the match algorithm for the build phrase

jimklimov commented on README.md:
```diff
@@ -42,24 +42,25 @@ Select *Git* then configure:
-- **Stash Credentials**: Select or Add the login username/password for the Stash REST API Host (NOT ssh key)
-- **Project**: abbreviated project code. eg: *PRJ* or *~user*
-- **RepositoryName**: eg: *Repo*
+- **Cron schedule**: How often to poll, e.g. every 2 minute: `H/2 * * * *`
```
2 minute*s*

jimklimov commented on README.md:
```diff
@@ -42,24 +42,25 @@ Select *Git* then configure:
+- **Cancel outdated jobs**: Cancel all jobs in the queue for the same PR.
+- **Phrase to disable builds**: Don't build the PR if the specified phrase has been posted in a PR comment. Default: *NO TEST*
+- **Only build if asked with the build phrase**: Only trigger the build when the build phrase has been posted.
+- **Phrase to request a build**: Build the PR if the specified phrase has been posted as a PR comment in Stash, even if the PR has not changed. Default: *test this please*
```
Maybe emphasize that this allows to `(Re-)Build the PR`? And maybe clarify that this can be useful if e.g. earlier tests failed due to external circumstances not related to the codebase in question. And also that direct "Rebuild" of such a job in Jenkins GUI won't work (it uses expansion of variables set by the plugin, and undefined in a native Rebuild).

jakub-bochenski replied:
> And also that direct "Rebuild" of such a job in Jenkins GUI won't work (it uses expansion of variables set by the plugin, and undefined in a native Rebuild).

I think if you use the https://wiki.jenkins.io/display/JENKINS/Rebuild+Plugin it would work. If it doesn't it should -- haven't actually tried it with this stash plugin, but it does work e.g. for the gitlab PR plugin

jimklimov commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly:
```diff
@@ -37,19 +37,19 @@
       <f:checkbox default="false"/>
     </f:entry>
-    <f:entry title="CI Build Phrases" field="ciBuildPhrases">
+    <f:entry title="Phrase to request a build" field="ciBuildPhrases">
```
Maybe `Phrase to request a (re-)build` as detailed in a comment for the README?

jimklimov commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-ciBuildPhrases.html:
```diff
@@ -0,0 +1,10 @@
+
+    The phrase should be a substring of the comment. The case is ignored.
+    For instance, "Test this please, Jenkins!" would match the default
+    setting.
```
Should here be a trailing `<br/>` ?

proski replied:
Yes, adding it.

jimklimov commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-projectCode.html:
```diff
@@ -0,0 +1,3 @@
+<div>
+    Project key (abbreviated name) in Stash, e.g. <em>PRJ</em>
```
.. `or <em>~username</em>` ?

proski replied:
Done, good point!

jimklimov commented on src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-stashHost.html:
```diff
@@ -0,0 +1,6 @@
+<div>
+    The http or https URL of the Stash server REST API.
```
Should here be a trailing `<br/>` before the NOTE?

proski replied:
Done

jimklimov commented on README.md:
```diff
@@ -42,24 +42,25 @@ Select *Git* then configure:
-- **Project**: abbreviated project code. eg: *PRJ* or *~user*
-- **RepositoryName**: eg: *Repo*
+- **Cron schedule**: How often to poll, e.g. every 2 minute: `H/2 * * * *`
+- **Stash URL**: The *http* or *https* URL of the Stash REST API (*NOT* ssh), e.g. *https://stash.example.com/*
```
General question for numerous places in the texts: shouldn't HTTP, HTTPS, SSH be upper-cased (they are abbreviations after all)? Especially since SSL was upper-cased by this PR ;)

proski replied:
I believe it was possible to select an ssh key for the Stash account, which was confusing. That required extra clarifications. The issue with the credential selection was fixed in version 1.7.0, but the clarifications are still there. We could tone them down slightly.

I'd rather remove reverences to http and https completely. I'm also thinking about removing all references to REST. No need to add an abbreviation that doesn't help understanding the text. Stash API is an API to Stash, that's it.

Standards use `HTTP`, but `http` looks better to me, as I see the lowercase version in URLs all the time.

I'm leaving that part as is for now.

jakub-bochenski replied:
> I'd rather remove reverences to http and https completely. I'm also thinking about removing all references to REST. No need to add an abbreviation that doesn't help understanding the text. Stash API is an API to Stash, that's it.

I'm not sure, I still find it useful to have a clarification that it's a HTTP URL and not and SSH Clone URL. On the other hand you can still use HTTP Clone URL. 

> Standards use HTTP, but http looks better to me, as I see the lowercase version in URLs all the time.

Well, HTTP is the abbreviated name of the protocol. `http` is an URI/URL scheme value

jakub-bochenski replied:
Let's get back to this point later as not to block the other changes

jimklimov reviewed (approved):
Generally looks OK and can be merged as is; however I posted some comments about clarifying the documentation better for users, and about some potential issues in HTML markup of some of the help files.

jimklimov commented:
Also as a note for posterity, the Wiki page was recently edited to remove duplicate effort of maintaining configuration details, and to refer to the README on github instead.

jakub-bochenski commented:
I'd rather wait for final version if @proski  can respond soon

proski committed:
Addressed PR comments

proski commented:
Force-push only due to rebasing. The latest changes are only in the last commit.

proski commented:
Using the same email address in all commits.
